File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.go
func lengthOfLongestSubstring(s string) int {
	charSet := make(map[rune]bool)
	left := 0
	maxLength := 0

	for i, r := range s {
		for charSet[r] {
			delete(charSet, rune(s[left]))
			left++
		}
		charSet[r] = true

		if i-left+1 > maxLength {
			maxLength = i - left + 1
		}
	}

	return maxLength
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.ts
function lengthOfLongestSubstring(s: string): number {
   let left = 0;
   let right = 0;
   let maxLength = 0;
   let track = new Set();
   while (right < s.length) {
    if (track.has(s[right])) {
        maxLength = maxLength > track.size ? maxLength : track.size;
        while( track.has(s[right]) ) {
            track.delete(s[left]);
            left++;
        }
    } 
    track.add(s[right]);
    right++
   }

    maxLength = maxLength > track.size ? maxLength : track.size;
    return maxLength;
};

